spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/startup/StartupTimeMetrics.java

```diff
@@ -0,0 +1,126 @@
+ * limitations under the License.
+ */
+
+package org.springframework.boot.actuate.metrics.startup;
+
+import java.lang.management.RuntimeMXBean;
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.TimeGauge;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.boot.context.event.ApplicationStartedEvent;
+import org.springframework.context.ApplicationEvent;
+import org.springframework.context.event.SmartApplicationListener;
+
+/**
+ * Binds application startup metrics in response to {@link ApplicationStartedEvent} and
+ * {@link ApplicationReadyEvent}.
+ *
+ * @author Chris Bono
+ * @since 2.6.0
+ */
+public class StartupTimeMetrics implements SmartApplicationListener {
+
+	private final MeterRegistry meterRegistry;
+
+	private final RuntimeMXBean runtimeMXBean;
+
+	private final String applicationStartedTimeMetricName;
+
+	private final String applicationReadyTimeMetricName;
+
+	private final String applicationReadyJvmTimeMetricName;
+
+	private final Iterable<Tag> tags;
+
+	public StartupTimeMetrics(MeterRegistry meterRegistry, RuntimeMXBean runtimeMXBean,
+			String applicationStartedTimeMetricName, String applicationReadyTimeMetricName,
+			String applicationReadyJvmTimeMetricName) {
+		this(meterRegistry, runtimeMXBean, Collections.emptyList(), applicationStartedTimeMetricName,
+				applicationReadyTimeMetricName, applicationReadyJvmTimeMetricName);
+	}
+
+	public StartupTimeMetrics(MeterRegistry meterRegistry, RuntimeMXBean runtimeMXBean, Iterable<Tag> tags,
+			String applicationStartedTimeMetricName, String applicationReadyTimeMetricName,
+			String applicationReadyJvmTimeMetricName) {
+		this.meterRegistry = meterRegistry;
+		this.runtimeMXBean = runtimeMXBean;
+		this.tags = (tags != null) ? tags : Collections.emptyList();
+		this.applicationStartedTimeMetricName = applicationStartedTimeMetricName;
+		this.applicationReadyTimeMetricName = applicationReadyTimeMetricName;
+		this.applicationReadyJvmTimeMetricName = applicationReadyJvmTimeMetricName;
+	}
+
+	@Override
+	public boolean supportsEventType(Class<? extends ApplicationEvent> eventType) {
+		return ApplicationStartedEvent.class.isAssignableFrom(eventType)
+				|| ApplicationReadyEvent.class.isAssignableFrom(eventType);
+	}
+
+	@Override
+	public void onApplicationEvent(ApplicationEvent event) {
+		if (event instanceof ApplicationStartedEvent) {
+			onApplicationStarted((ApplicationStartedEvent) event);
+		}
+		if (event instanceof ApplicationReadyEvent) {
+			onApplicationReady((ApplicationReadyEvent) event);
+		}
+	}
+
+	private void onApplicationStarted(ApplicationStartedEvent event) {
+		if (event.getStartupTime() == null) {
+			return;
+		}
+		TimeGauge
```

Comment:
Wdyt about the auto-formatting on this one? It's not the easiest on the eyes.

Reply:
I'm not too worried about it. We might try some "extract variable/method" refactoring when we merge.